### ISSUE TITLE
vespa-proton-bin | Assertion `_current_index_id < ISourceSelector::SOURCE_LIMIT' failed.

### ISSUE THREAD

**Issue by jblankfeld** [vespa-engine/vespa#13459]
Hi,
I am trying to set up a Vespa cluster with dozens of nodes and billions of documents.
There are 3 nodes that fail to start because searchnode process fails to start.
This happened after Vespa finished building the document database (I do vespa-proton-cmd --local getProtonStatus) and the node is permanently down after that.

```
[2020-06-02 14:22:05.799] WARNING : searchnode       proton.vespalib.util.process_memory_stats	We failed to find 2 consecutive samples that where similar with epsilon of 1048576.\nSmallest is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=73349718016, _anonymous_rss=64711692288, _mappings_count=23137',\n median is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=73915949056, _anonymous_rss=65228787712, _mappings_count=23138',\n largest is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=74087915520, _anonymous_rss=65386004480, _mappings_count=23136'
[2020-06-02 14:22:17.971] WARNING : container        Container.com.yahoo.search.dispatch.searchcluster.SearchCluster	Coverage of group 0 is only 57/60 (requires 60) (13012352776/13012352776 active docs) Failed nodes are:\nsearch node key = 68 hostname = xxxxxxxx path = 56 in group 0 statusIsKnown = true working = false activeDocs = 0\nsearch node key = 69 hostname = xxxxxxxx path = 57 in group 0 statusIsKnown = true working = false activeDocs = 0\nsearch node key = 70 hostname = xxxxxxxx path = 58 in group 0 statusIsKnown = true working = false activeDocs = 0
[2020-06-02 14:22:34.518] WARNING : searchnode       proton.vespalib.util.process_memory_stats	We failed to find 2 consecutive samples that where similar with epsilon of 1048576.\nSmallest is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=74796752896, _anonymous_rss=66096906240, _mappings_count=23136',\n median is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=75769831424, _anonymous_rss=66301804544, _mappings_count=23138',\n largest is '_mapped_virt=1373291925504, _mapped_rss=83451904, _anonymous_virt=77067968512, _anonymous_rss=67812179968, _mappings_count=23136'
[2020-06-02 14:22:38.268] INFO    : searchnode       proton.searchcorespi.index.indexmaintainer	Making new schema. Id = 901. Serial num = 211322319
[2020-06-02 14:22:38.398] WARNING : searchnode       stderr	vespa-proton-bin: /builddir/build/BUILD/vespa-7.193.27/searchcorespi/src/vespa/searchcorespi/index/indexmaintainer.cpp:775: void searchcorespi::index::IndexMaintainer::doneSetSchema(searchcorespi::index::IndexMaintainer::SetSchemaArgs&, searchcorespi::index::IMemoryIndex::SP&): Assertion `_current_index_id < ISourceSelector::SOURCE_LIMIT' failed.
```

Do you have an idea how to restore those nodes ? Can I recover the data at least ?

**Comment by jblankfeld** [comment]
I now have 15 failed nodes and log message is slightly different this time:
```
[2020-06-03 05:08:24.722] WARNING : searchnode       stderr	vespa-proton-bin: /builddir/build/BUILD/vespa-7.193.27/searchcorespi/src/vespa/searchcorespi/index/indexmaintainer.cpp:480: bool searchcorespi::index::IndexMaintainer::doneInitFlush(searchcorespi::index::IndexMaintainer::FlushArgs*, searchcorespi::index::IMemoryIndex::SP*): Assertion `_current_index_id < ISourceSelector::SOURCE_LIMIT' failed.
```

**Comment by baldersheim** [comment]
This we have never seen before. I think this is a sanity check that verifies that we stay below 250 index generations. I have never sen this number higher than 3 or 4. This mean that you have fed data faster than it is able to fusion them.

We have not added back pressure here as this was considered impossible.

What kind of hardware are you running on ? Number of cores, amount of memory, what kind of disks ?

**Comment by jblankfeld** [comment]
Hi,
We currently have 60 nodes with 40 CPU, 128GB RAM and 3.5TB HDD for each.
I use 50 threads to feed to Vespa using Spark and the Vespa HTTP Client library.
It's true there are many index.flush folders in 
```
ls /opt/vespa/var/db/vespa/search/cluster.mycluster/n6/documents/mycluster/0.ready/index/
index.flush.1561  index.flush.1584  index.flush.1607  index.flush.1630  index.flush.1653  index.flush.1676  index.flush.1699  index.flush.1722  index.flush.1745  index.flush.1768  index.flush.1791
index.flush.1562  index.flush.1585  index.flush.1608  index.flush.1631
...
```
I am going to tune https://docs.vespa.ai/documentation/proton.html#memory-index-flush and feed at a slower rate. Do you have other pieces of advice ?

**Comment by baldersheim** [comment]
Since I do not know how much memory is consumed by attributes I can not give a perfect advice, but I think that setting maxmemorygain to 16G for both total and component in https://docs.vespa.ai/documentation/reference/services-content.html#flushstrategy is a safe choice. The default of 1G is way too small for large nodes.

I also recommend to do both feed and query benchmarking on a single node first. It is a lot faster to iterate to achieve proper tuning with a single node, than with multiple.

**Comment by jblankfeld** [comment]
Thanks for this tip, this helped a lot.
I still had high number of memory flushes during it but it stayed under the hard limit so it went well.

**Comment by jobergum** [comment]
Thanks for coming back to us on this @jblankfeld, it would be interesting to know more about your setup and use case if you can share.

**Comment by jblankfeld** [comment]
Of course, my organization is doing a web search experiment, with roughly ~12 billions of documents, ~40 various fields (static scores, variant of stemmed and not stemmed title, h1, metas, anchors etc...) and the hardware setup I described above.
I hit this issue a couple times, including when I added a node to the existing cluster.
The flushes were less than 512MiB and I eventually reached the limit set to 250.
But I managed to workaround it thanks to the suggestion from @baldersheim, I also set feeding.concurrency to 1.0 and flushes went up to ~3GiB which solved the issue.
